flare/algorithm_zoo/distributional_rl_algorithms.py

```diff
@@ -0,0 +1,234 @@
+        :return: Tensor (batch_size x num_atoms). Q value distribution for an
+            action.
+        """
+        one_hot_action = comf.one_hot(
+            action.squeeze(-1), q_distributions.size()[1])
+        one_hot_action = one_hot_action.unsqueeze(1)
+        q_distribution = torch.matmul(one_hot_action, q_distributions)
+        return q_distribution.squeeze(1)
+
+    def learn(self, inputs, next_inputs, states, next_states, next_alive,
+              actions, next_actions, rewards):
+
+        if self.update_ref_interval \
+                and self.total_batches % self.update_ref_interval == 0:
+            ## copy parameters from self.model to self.ref_model
+            self.ref_model.load_state_dict(self.model.state_dict())
+        self.total_batches += 1
+
+        action = actions["action"]
+        reward = rewards["reward"]
+
+        values, states_update = self.get_current_value(inputs, states)
+        q_lists = values["q_value_list"]
+
+        with torch.no_grad():
+            next_values, next_states_update = self.get_next_value(next_inputs,
+                                                                  next_states)
+            next_q_lists = self.check_alive(next_values, next_alive)
+            next_expected_q_values = next_values["q_value"]
+            _, next_action = next_expected_q_values.max(-1)
+            next_action = next_action.unsqueeze(-1)
+
+        assert q_lists.size()[:2] == next_q_lists.size()[:2]
+
+        q_list = self.select_q_distribution(q_lists, action)
+        next_q_list = self.select_q_distribution(next_q_lists, next_action)
+
+        cost = self.get_cost(q_list, next_q_list, reward, values, next_values)
+        return dict(cost=cost), states_update, next_states_update
+
+    def get_current_value(self, inputs, states):
+        return self.model.value(inputs, states)
+
+    def get_next_value(self, next_inputs, next_states):
+        return self.ref_model.value(next_inputs, next_states)
+
+    def check_alive(self, next_values, next_alive):
+        pass
+
+    def get_cost(self, q_list, next_q_list, reward, values, next_values):
+        pass
+
+
+class C51(DistributionalAlgorithm):
+    def __init__(self,
+                 model,
+                 gpu_id=-1,
+                 discount_factor=0.99,
+                 exploration_end_steps=0,
+                 exploration_end_rate=0.1,
+                 update_ref_interval=100):
+        super(C51, self).__init__(model, gpu_id, discount_factor,
+                                  exploration_end_steps, exploration_end_rate,
+                                  update_ref_interval)
+        dead_dist = [0.] * self.model.bins
+        dead_dist[0] = 1.
+        self.dead_dist = torch.tensor(dead_dist)
+        self.float_vmax = torch.FloatTensor([model.vmax])
+        self.float_vmin = torch.FloatTensor([model.vmin])
+
+    def check_alive(self, next_values, next_alive):
+        ## if not alive, Q value is the minimum.
```

Comment:
Is the minimum guaranteed to be zero? When the agent is dead, it should receive zero rewards afterwards.